http2-tests/src/test/scala/org/apache/pekko/http/impl/engine/http2/Http2SpecWithMaterializer.scala

```diff
@@ -0,0 +1,138 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * license agreements; and to You under the Apache License, version 2.0:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * This file is part of the Apache Pekko project, which was derived from Akka.
+ */
+
+/*
+ * Copyright (C) 2018-2022 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package org.apache.pekko.http.impl.engine.http2
+
+import org.apache.pekko
+import pekko.NotUsed
+import pekko.event.Logging
+import pekko.http.impl.engine.http2.FrameEvent._
+import pekko.http.impl.engine.server.ServerTerminator
+import pekko.http.impl.engine.ws.ByteStringSinkProbe
+import pekko.http.impl.util.PekkoSpecWithMaterializer
+import pekko.http.impl.util.LogByteStringTools
+import pekko.http.scaladsl.Http
+import pekko.http.scaladsl.model._
+import pekko.http.scaladsl.settings.ServerSettings
+import pekko.stream.Attributes
+import pekko.stream.Attributes.LogLevels
+import pekko.stream.scaladsl.{ BidiFlow, Flow, Keep, Sink, Source }
+import pekko.stream.testkit.TestPublisher.Probe
+import pekko.stream.testkit.scaladsl.StreamTestKit
+import pekko.stream.testkit.{ TestPublisher, TestSubscriber }
+import pekko.util.ByteString
+
+import scala.collection.immutable
+import scala.concurrent.{ ExecutionContext, Future }
+
+abstract class Http2SpecWithMaterializer(configOverrides: String) extends PekkoSpecWithMaterializer(configOverrides) {
+  implicit class InWithStoppedStages(name: String) {
+    def inAssertAllStagesStopped(runTest: => TestSetup) =
+      name in StreamTestKit.assertAllStagesStopped {
+        val setup = runTest
+
+        // force connection to shutdown (in case it is an invalid state)
+        setup.network.fromNet.sendError(new RuntimeException)
+        setup.network.toNet.cancel()
+
+        // and then assert that all stages, substreams in particular, are stopped
+      }
+  }
+
+  protected /* To make ByteFlag warnings go away */ abstract class TestSetupWithoutHandshake {
```

Comment:
The comment after `protected` doesn't seem elegant

Reply:
https://github.com/apache/incubator-pekko-http/pull/344/files/a58637af2969f314cd900b44cb87523dc60e61fd#diff-1bf6a966e29f3f599b7126edc971a2eb45087a0f37ff3b578389f38f257424a5L1809

Reply:
yeah, Should we keep it orgin? If it's not necessary, I still hope to be able to modify it.